### PR TITLE
[Model] Mamba2: Bit-identical prefill and decode under batch-invariant mode

### PR DESCRIPTION
**Pull request by NolenLiang**
## Purpose

Second half of the split of this PR: #57539 adds the exact-replay SSD step and its metadata; this PR wires them into the Mamba2 backend. Tracked in #55524.

Under `VLLM_BATCH_INVARIANT=1`, a Mamba2 layer should produce the same bits for a token whether it is computed in prefill, in a chunked-prefill continuation or in decode. Today prefill runs the chunked scan and decode the single-token update, which round differently, so a token's logprob depends on how the scheduler split the sequence. That defeats the purpose of the mode for RL, where a trainer rescores the rollout through prefill.

With this PR, in batch-invariant mode:

- The SSM state is kept in fp32 and the layer state gains three per-sequence buffers with the inputs of the current partial chunk (x, dt, B), through the shared state shape and dtype calculators, so every model built on `MambaMixer2` gets it without per-model changes.
- The Mamba2 attention metadata builder produces the replay metadata for the prefill and decode rows of each step, and the mixer runs the replayed step from #57539 instead of the chunked scan and the single-token update.
- Unsupported settings are rejected at startup: prefix caching on the Mamba2 layers, PP > 1, speculative decoding, micro-batching, KV connectors and the non-Triton mamba backend. Tensor parallelism works; the gated RMSNorm's TP path for a single group reduces per-rank means in fp32, because batch-invariant mode overrides `mean` but not `sum` (#54993 touches the same lines).
- The Mamba2 layers run outside CUDA graphs in this mode; #55905 brings them back.

Default mode is untouched. Costs: each decode step runs the chunked-scan kernels over the partial chunk instead of the single-token update, the per-layer state grows from two tensors to five (about 43 MiB per sequence for a 27-layer, 32-head model at chunk size 256), and decode throughput on small models drops to roughly a tenth. #55905 recovers most of that, and #55895 pins the kernels' tiles so the bits also match across processes.

Not a duplicate: #54993 aligns the scheduler's chunking and replays the generated history through the SSU kernel on recovery, so it gives decode == recompute while decode stays on SSU. This PR gives decode == single-shot prefill.

## Test Plan

```bash
pytest tests/v1/attention/test_mamba2_batch_invariant.py
pytest tests/v1/determinism/test_mamba2_batch_invariance.py
```

Engine-level oracle, out of tree: tokens are forced through a logits processor so every run realizes the same trajectory, then each generated token's raw logprob is compared bit for bit with its raw logprob in a single-shot prefill of prompt plus generation.

## Test Result

GB200, TP1 unless noted, on the final commit.

- Unit tests: 7 cases for the state layout and the startup checks, plus the engine-level determinism test.
- mamba2-130m, 8 concurrent requests, prompts of 37 to 1000 tokens, 60 to 300 generated tokens: 0 of 3666 prefill and 0 of 1190 decode positions differ, with and without `max_num_batched_tokens=100`, CUDA graphs on. With the mode off, 1182 of 1190 differ.
- granite-4.0-h-350m with `num_gpu_blocks_override=33`, so one request is preempted and recomputed: 0 of 1859 and 0 of 5750, with CUDA graphs and eager. Mode off: 5746 of 5750.
- Nemotron-H-8B-Base-8K: 0 of 1859 and 0 of 5750 at TP1 and at TP2.
- An earlier revision of this change also passed the determinism test on H100 NVL, posted in this thread.
- pre-commit passes on the changed files.

## AI assistance

AI tools assisted with the implementation, tests and documentation. I reviewed every changed line and ran the tests above.

### CODE REVIEW

File: vllm/model_executor/layers/mamba/exact_replay.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    if augmented:
+        buffered_slot = slots64[meta.buffered_seq]
+        x_aug = x.new_empty((n_aug, nheads, head_dim))
+        dt_aug = dt.new_empty((n_aug, nheads))
+        B_aug = B.new_empty((n_aug, ngroups, dstate))
+        C_aug = C.new_empty((n_aug, ngroups, dstate))
+        x_aug[meta.buffered_dst] = buffers.x[buffered_slot, meta.buffered_pos]
+        dt_aug[meta.buffered_dst] = buffers.dt[buffered_slot, meta.buffered_pos]
+        B_aug[meta.buffered_dst] = buffers.B[buffered_slot, meta.buffered_pos]
+        C_aug[meta.buffered_dst] = buffers.C[buffered_slot, meta.buffered_pos]
```
Comment (jzakrzew) [inline vllm/model_executor/layers/mamba/exact_replay.py]: You don't really need historical C for decode

File: vllm/model_executor/layers/mamba/exact_replay.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        B_aug[meta.step_dst] = B
+        C_aug[meta.step_dst] = C
+        out_aug = torch.empty_like(x_aug)
+    else:
+        x_aug, dt_aug, B_aug, C_aug, out_aug = x, dt, B, C, out
+
+    initial_states = torch.where(
+        meta.has_boundary_state[:, None, None, None], ssm_state[slots], 0
+    )
+    states = mamba_chunk_scan_combined_varlen(
```
Comment (jzakrzew) [inline vllm/model_executor/layers/mamba/exact_replay.py:110]: not materializing the whole state on each decode step is a huge optimization opportunity here. That said, this would require specialized kernels, so it would probably be too big for one PR.

File: vllm/model_executor/layers/mamba/exact_replay.py (hunk 1 above)
Comment (NolenLiang) [reply vllm/model_executor/layers/mamba/exact_replay.py]: Right, thanks. C only enters a token's own output row and the re-fed rows are discarded, so the buffer is gone: the state is now (conv, ssm, x, dt, B) and the re-fed positions get zeros for C. Tests and the engine-level checks are unchanged (all bitwise equal).

File: vllm/model_executor/layers/mamba/exact_replay.py (hunk 2 above)
Comment (NolenLiang) [reply vllm/model_executor/layers/mamba/exact_replay.py:110]: Agreed. This PR recomputes the whole partial chunk and materializes the boundary state on every step; the row-gated version (only the consumed row and, on a chunk boundary, the new state) needs modified SSD kernels and is the natural follow-up once the semantics here are settled. I will note it in the RFC as the next step.

**Comment by NolenLiang** [comment]
Pushed eb6591e5e with three changes since the first review:

1. C is no longer buffered; the state is (conv, ssm, x, dt, B).
2. Tensor parallelism is allowed now. The only TP blocker was the gated RMSNorm's TP > 1 path for `n_groups == 1`, which reduced per-rank sums with `torch.sum`; its row reduction order depends on the token count, so decode diverged from prefill at data-dependent steps. In batch-invariant mode that path now reduces per-rank means (`mean` is covered by the batch-invariant overrides, `sum` is not). Verified at TP=2 on mamba2-130m, granite-4.0-h-350m under preemption and Nemotron-H-8B, all bitwise equal to single-shot prefill; a dense model at TP=2 was already equal, as a control.
3. Description updated with measured end-to-end costs (FlashAttention pinned to v2 on both sides) and the Nemotron-H-8B results.

**Comment by yuvalluria** [comment]
Tested on H100 NVL (SM90) — pure Python overlay of PR changes onto nightly `vllm/vllm-openai:nightly` (0.28.1rc1.dev580+g385dce36b):

\`\`\`
NVIDIA H100 NVL, 9.0, 95830 MiB
Torch: 2.13.0+cu130 | CUDA: 13.0
vLLM: 0.28.1rc1.dev580+g385dce36b

============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-9.1.1, pluggy-1.6.0 -- /usr/bin/python3
rootdir: /tmp/det
plugins: anyio-4.15.1
collecting ... collected 1 item

test_mamba2_batch_invariance.py::test_mamba2_decode_logprobs_match_prefill_logprobs PASSED [100%]

================== 1 passed, 16 warnings in 108.05s (0:01:48) ==================
\`\`\`

**Comment by NolenLiang** [comment]
@yuvalluria  Thanks a lot for running this on H100, that's the first SM90 data point (I only have GB200 here). If the box is still handy, the stacked #55905 is where an SM90 run would help most: it adds the row-gated decode kernels with pinned tiles and full CUDA graphs. `tests/kernels/mamba/test_mamba2_ssd_emit.py` plus the same determinism test cover it.

**Comment by NolenLiang** [comment]
@tdoublep @yewentao256 this is now out of draft, together with the stacked #55905 and #56244. It makes Mamba2 prefill, chunked prefill and decode bitwise identical under `VLLM_BATCH_INVARIANT=1` by replaying the partial chunk from an fp32 boundary state (design in #55524). Happy to split or reorder anything that makes the review easier.

**Comment by NolenLiang** [comment]
@tomeras91 as one of the owners of the Mamba code, could you take a look? This PR (with #55895 as its base and #55905 / #56244 stacked on top) is the batch-invariant Mamba2 work tracked in #55524; the stack is out of draft.

**Review by yewentao256 — COMMENTED** [review]
Please shrink the diff idealy <400 LOC.

What's the context of need to support mamba2?

**Comment by NolenLiang** [comment]
Fair point on the size, I'll split it. Two PRs, one with the exact-replay ops, the metadata and their unit tests, one with the state layout and the model/backend wiring plus the engine tests. The TP norm variance change and the docs come out of this PR (#54993 touches the same norm code). Does that split work for you?

As for why Mamba2, it's not a new model. Nemotron-H and the other hybrids already run, but batch-invariant mode doesn't actually hold for them. Prefill goes through the chunked SSD kernels and decode through selective_state_update, and those two aren't bit-identical, so a token's logprob depends on where the scheduler split the sequence. For RL (NeMo RL) the rollout gets rescored through prefill on the trainer side, and that's where it breaks. Numbers are in #55524.

**Comment by NolenLiang** [comment]
> Please shrink the diff idealy <400 LOC.
> 
> What's the context of need to support mamba2?

Fair point on the size, I'll split it. Two PRs, one with the exact-replay ops, the metadata and their unit tests, one with the state layout and the model/backend wiring plus the engine tests. The TP norm variance change and the docs come out of this PR (#54993 touches the same norm code). Does that split work for you?

As for why Mamba2, it's not a new model. Nemotron-H and the other hybrids already run, but batch-invariant mode doesn't actually hold for them. Prefill goes through the chunked SSD kernels and decode through selective_state_update, and those two aren't bit-identical, so a token's logprob depends on where the scheduler split the sequence. For RL (NeMo RL) the rollout gets rescored through prefill on the trainer side, and that's where it breaks. Numbers are in #55524.

**Comment by NolenLiang** [comment]
Split as discussed. The replayed SSD step and its metadata are now #57539; this PR keeps the state layout, the backend wiring and the engine tests, rebased on main. The description is updated and the numbers there are from the new head.

**Comment by NolenLiang** [comment]
A correction to my earlier reply. The TP norm variance change is still in this PR. It is seven lines in the mixer and TP2 needs it to stay bit-identical, so I kept it. The docs are out as said. If #54993 lands first I will drop it here.

**Comment by yuvalluria** [comment]
Happy to run H100 NVL validation on this stack once it's rebased. We have an H100 NVL cluster running the nightly image and just used the same setup to validate PR #57280 (layernorm_guard batch invariance). Once #55627, #55905, and #56244 are rebased and ready, ping here and I'll post raw test output.

**Comment by NolenLiang** [comment]
Thanks, that would be useful. The stack is rebased on main now, #57539 sits under this PR and #55905 and #56244 are stacked on top. Running `pytest tests/v1/determinism/test_mamba2_batch_invariance.py` on the head of #56244 covers the whole stack, the conftest switches the mode on. Raw output here is welcome.
